### ISSUE TITLE
wasm: no longer compile my project

### ISSUE THREAD

**Issue by ryuukk** [ldc-developers/ldc#4324]
```
OS: Windows 10 x64
$ ldc2 --version
LDC - the LLVM D compiler (1.31.0):
  based on DMD v2.101.2 and LLVM 14.0.3
  built with LDC - the LLVM D compiler (1.31.0)
  Default target: x86_64-pc-windows-msvc
  Host CPU: skylake
  http://dlang.org - http://wiki.dlang.org/LDC
```
Hello,

My project no longer build for wasm, i'm not sure i understand why, i made sure no code import modules from libc that isn't in its proper version block for platform specific code

I get the following:

https://gist.github.com/ryuukk/c14e8152b1071baa2ce6fba5f9d11290

How should i debug this?

Why doesn't the compiler tell me what imports it?

Why does the compiler compile code that i supposedly do not use/import?

**Comment by kinke** [comment]
This is a failure to compile the `dub_platform_probe` module, something dub does implicitly to probe the compiler for the target platform. See https://github.com/dlang/dub/blob/9b15ad07e1f0731cd95c72e9d2119ebdac3d771d/source/dub/compilers/utils.d#L317-L363 for its source; seems like it's not compatible with what you set up in your `ldc2.conf`. Running dub with `--vverbose` (IIRC) might show more details.

**Comment by kinke** [comment]
You'd need to get hold of the temp .d file being compiled (with `-mtriple=…` for wasm) if dub really doesn't print it in its logs (and doesn't even show the cmdline apparently); that'd be some `dub_platform_probe_089b65e3_594f_4193_a820_02e334021449.d` file somewhere in your filesystem (maybe still there, maybe already removed from dub). This file then includes regular druntime/Phobos (inherited `post-switches`), not your modified version with the wasm fixes.

**Comment by kinke** [comment]
Yeah, you'll only be able to use dub if that file can be compiled for your target (with `-o-`, i.e., no codegen required). And looks like that regressed with D v2.101; I guess due to new templated druntime hooks.

**Comment by kinke** [comment]
> How should i debug this?

I've used `-vcg-ast` to dump the produced AST (without `-mtriple=wasm32-…`). That writes a `.cg` file next to the source file. And the problem is indeed a new array-append hook:

v1.30 AST:
```D
module dub_platform_probe;
import object;
enum toString(int v) = v.stringof;
string stringArray(string[] ary)
{
	string res = null;
	{
		scope string[] __r3 = ary[];
		ulong __key2 = 0LU;
		for (; __key2 < __r3.length; __key2 += 1LU)
		{
			string e = __r3[__key2];
			ulong i = __key2;
			if (i)
				res ~= ", ";
			res ~= '"' ~ e ~ '"';
		}
	}
	return res;
}
pragma (msg, "__dub_probe_begin__\n{\n  \"compiler\": \"" ~ determineCompiler() ~ "\",\n  \"frontendVersion\": 2100,\n  \"compilerVendor\": \"LDC\",\n  \"platform\": [\n    " ~ stringArray(determinePlatform()) ~ "\n  ],\n  \"architecture\": [\n    " ~ stringArray(determineArchitecture()) ~ "\n   ],\n}\n__dub_probe_end__");
string[] determinePlatform()
{
	string[] ret = null;
	ret ~= "linux";
	ret ~= "posix";
	return ret;
}
string[] determineArchitecture()
{
	string[] ret = null;
	ret ~= "x86_64";
	return ret;
}
string determineCompiler()
{
	return "ldc";
}
toString!2100
{
	enum string toString = "2100";

}
```

v1.31 AST:
```D
module dub_platform_probe;
import object;
enum toString(int v) = v.stringof;
string stringArray(string[] ary)
{
	string res = null;
	{
		scope string[] __r3 = ary[];
		ulong __key2 = 0LU;
		for (; __key2 < __r3.length; __key2 += 1LU)
		{
			string e = __r3[__key2];
			ulong i = __key2;
			if (i)
				_d_arrayappendT(res, ", ");
			_d_arrayappendT(res, '"' ~ e ~ '"');
		}
	}
	return res;
}
pragma (msg, "__dub_probe_begin__\n{\n  \"compiler\": \"" ~ determineCompiler() ~ "\",\n  \"frontendVersion\": 2101,\n  \"compilerVendor\": \"LDC\",\n  \"platform\": [\n    " ~ stringArray(determinePlatform()) ~ "\n  ],\n  \"architecture\": [\n    " ~ stringArray(determineArchitecture()) ~ "\n   ],\n}\n__dub_probe_end__");
string[] determinePlatform()
{
	string[] ret = null;
	(string __appendtmp7 = "linux";) , (_d_arrayappendcTX(ret, 1LU) , ret[ret.length - 1LU] = __appendtmp7 , ret);
	(string __appendtmp8 = "posix";) , (_d_arrayappendcTX(ret, 1LU) , ret[ret.length - 1LU] = __appendtmp8 , ret);
	return ret;
}
string[] determineArchitecture()
{
	string[] ret = null;
	(string __appendtmp9 = "x86_64";) , (_d_arrayappendcTX(ret, 1LU) , ret[ret.length - 1LU] = __appendtmp9 , ret);
	return ret;
}
string determineCompiler()
{
	return "ldc";
}
toString!2101
{
	enum string toString = "2101";

}
_d_arrayappendT!(string, immutable(char))
{
	pure nothrow ref @trusted string _d_arrayappendT(return ref scope string x, scope string y)
	{
		import core.stdc.string : memcpy;
		import core.internal.traits : hasElaborateCopyConstructor, Unqual;
		import core.lifetime : copyEmplace;
		enum bool hasPostblit = false;
		ulong length = x.length;
		_d_arrayappendcTX(x, y.length);
		if (y.length)
		{
			char* xptr = cast(char*)&x[length];
			immutable immutable(ulong) size = 1LU;
			memcpy(cast(void*)xptr, cast(const(void*))cast(char*)&y[0], y.length * 1LU);
		}
		return x;
	}

}
_d_arrayappendcTXImpl!(string, immutable(char))
{
	import core.internal.array.utils : _d_HookTraceImpl;
	private enum string errorMessage = "Cannot append to array if compiling without support for runtime type information!";
	pure nothrow ref @trusted string _d_arrayappendcTX(return ref scope string px, ulong n)
	{
		TypeInfo_Array ti = typeid(string);
		byte[] pxx = (cast(byte*)cast(immutable(char)*)px)[0..px.length];
		_d_arrayappendcTX(ti, pxx, n);
		px = (cast(immutable(char)*)cast(byte*)pxx)[0..pxx.length];
		return px;
	}
	alias _d_arrayappendcTXTrace = auto pure nothrow @trusted string _d_HookTraceImpl(string file, int line, string funcname, return ref scope string _param_3, ulong _param_4)
	{
		import core.internal.array.utils : gcStatsPure, accumulatePure;
		string name = "string";
		ulong currentlyAllocated = gcStatsPure().allocatedInCurrentThread;
		try
		{
			return _d_arrayappendcTX(_param_3, _param_4);
		}
		finally
		{
			ulong size = gcStatsPure().allocatedInCurrentThread - currentlyAllocated;
			if (size > 0LU)
				if (!accumulatePure(file, line, funcname, name, size))
				{
					assert(0);
				}
		}
	}
	;
}
_d_HookTraceImpl!(string, _d_arrayappendcTX, "Cannot append to array if compiling without support for runtime type information!")
{
	auto pure nothrow @trusted string _d_HookTraceImpl(string file, int line, string funcname, return ref scope string _param_3, ulong _param_4)
	{
		import core.internal.array.utils : gcStatsPure, accumulatePure;
		string name = "string";
		ulong currentlyAllocated = gcStatsPure().allocatedInCurrentThread;
		try
		{
			return _d_arrayappendcTX(_param_3, _param_4);
		}
		finally
		{
			ulong size = gcStatsPure().allocatedInCurrentThread - currentlyAllocated;
			if (size > 0LU)
				if (!accumulatePure(file, line, funcname, name, size))
				{
					assert(0);
				}
		}
	}

}
Parameters!(_d_arrayappendcTX)
{
	alias Parameters = (return ref scope string, ulong);

}
FunctionTypeOf!(_d_arrayappendcTX)
{
	alias pure nothrow ref @trusted string FunctionTypeOf(return ref scope string px, ulong n);

}
hasElaborateCopyConstructor!(immutable(char))
{
	enum bool hasElaborateCopyConstructor = false;

}
Unqual!(immutable(char), char)
{
	alias Unqual = char;

}
TraceHook!("string", "_d_arrayappendcTX")
{
	const const(char[]) TraceHook = "\n        import core.internal.array.utils : gcStatsPure, accumulatePure;\n\n        pragma(inline, false);\n        string name = `string`;\n\n        // FIXME: use rt.tracegc.accumulator when it is accessable in the future.\n        version (tracegc)\n    {\n\n            import core.stdc.stdio;\n\n            printf(\"%sTrace file = '%.*s' line = %d function = '%.*s' type = %.*s\\n\",\n            \"_d_arrayappendcTX\".ptr,\n                file.length, file.ptr,\n                line,\n                funcname.length, funcname.ptr,\n                name.length, name.ptr\n            );\n        }\n\n        ulong currentlyAllocated = gcStatsPure().allocatedInCurrentThread;\n\n        scope(exit)\n        {\n            ulong size = gcStatsPure().allocatedInCurrentThread - currentlyAllocated;\n            if (size > 0)\n                if (!accumulatePure(file, line, funcname, name, size)) {\n                    // This 'if' and 'assert' is needed to force the compiler to not remove the call to\n                    // `accumulatePure`. It really want to do that while optimizing as the function is\n                    // `pure` and it does not influence the result of this hook.\n\n                    // `accumulatePure` returns the value of `size`, which can never be zero due to the\n                    // previous 'if'. So this assert will never be triggered.\n                    assert(0);\n                }\n        }\n    ";

}
_d_arrayappendcTXImpl!(string[], string)
{
	import core.internal.array.utils : _d_HookTraceImpl;
	private enum string errorMessage = "Cannot append to array if compiling without support for runtime type information!";
	pure nothrow ref @trusted string[] _d_arrayappendcTX(return ref scope string[] px, ulong n)
	{
		TypeInfo_Array ti = typeid(string[]);
		byte[] pxx = (cast(byte*)cast(string*)px)[0..px.length];
		_d_arrayappendcTX(ti, pxx, n);
		px = (cast(string*)cast(byte*)pxx)[0..pxx.length];
		return px;
	}
	alias _d_arrayappendcTXTrace = auto pure nothrow @trusted string[] _d_HookTraceImpl(string file, int line, string funcname, return ref scope string[] _param_3, ulong _param_4)
	{
		import core.internal.array.utils : gcStatsPure, accumulatePure;
		string name = "string[]";
		ulong currentlyAllocated = gcStatsPure().allocatedInCurrentThread;
		try
		{
			return _d_arrayappendcTX(_param_3, _param_4);
		}
		finally
		{
			ulong size = gcStatsPure().allocatedInCurrentThread - currentlyAllocated;
			if (size > 0LU)
				if (!accumulatePure(file, line, funcname, name, size))
				{
					assert(0);
				}
		}
	}
	;
}
_d_HookTraceImpl!(string[], _d_arrayappendcTX, "Cannot append to array if compiling without support for runtime type information!")
{
	auto pure nothrow @trusted string[] _d_HookTraceImpl(string file, int line, string funcname, return ref scope string[] _param_3, ulong _param_4)
	{
		import core.internal.array.utils : gcStatsPure, accumulatePure;
		string name = "string[]";
		ulong currentlyAllocated = gcStatsPure().allocatedInCurrentThread;
		try
		{
			return _d_arrayappendcTX(_param_3, _param_4);
		}
		finally
		{
			ulong size = gcStatsPure().allocatedInCurrentThread - currentlyAllocated;
			if (size > 0LU)
				if (!accumulatePure(file, line, funcname, name, size))
				{
					assert(0);
				}
		}
	}

}
Parameters!(_d_arrayappendcTX)
{
	alias Parameters = (return ref scope string[], ulong);

}
FunctionTypeOf!(_d_arrayappendcTX)
{
	alias pure nothrow ref @trusted string[] FunctionTypeOf(return ref scope string[] px, ulong n);

}
TraceHook!("string[]", "_d_arrayappendcTX")
{
	const const(char[]) TraceHook = "\n        import core.internal.array.utils : gcStatsPure, accumulatePure;\n\n        pragma(inline, false);\n        string name = `string[]`;\n\n        // FIXME: use rt.tracegc.accumulator when it is accessable in the future.\n        version (tracegc)\n    {\n\n            import core.stdc.stdio;\n\n            printf(\"%sTrace file = '%.*s' line = %d function = '%.*s' type = %.*s\\n\",\n            \"_d_arrayappendcTX\".ptr,\n                file.length, file.ptr,\n                line,\n                funcname.length, funcname.ptr,\n                name.length, name.ptr\n            );\n        }\n\n        ulong currentlyAllocated = gcStatsPure().allocatedInCurrentThread;\n\n        scope(exit)\n        {\n            ulong size = gcStatsPure().allocatedInCurrentThread - currentlyAllocated;\n            if (size > 0)\n                if (!accumulatePure(file, line, funcname, name, size)) {\n                    // This 'if' and 'assert' is needed to force the compiler to not remove the call to\n                    // `accumulatePure`. It really want to do that while optimizing as the function is\n                    // `pure` and it does not influence the result of this hook.\n\n                    // `accumulatePure` returns the value of `size`, which can never be zero due to the\n                    // previous 'if'. So this assert will never be triggered.\n                    assert(0);\n                }\n        }\n    ";

}
```

**Comment by Geod24** [comment]
I never considered Wasm / other target with this. We should probably expose a way in dub to disable it / hardcode it.

**Comment by kinke** [comment]
I don't think that'd be easy, and the problem is much bigger than just dub's probe file - any string-concat for CTFE etc. doesn't work anymore in general for such targets with incomplete C bindings in druntime. If the problem here at least is only the `memcpy` import, then we have an `ldc.intrinsics.llvm_memcpy` alternative for LDC.

**Comment by kinke** [comment]
I've tested this - not with master, but with #4323. `memcpy` wasn't the problem, but all the `-profile=gc` extra stuff (not supported by LDC anyway) was. Applying my DMD PR linked above got the dub probe file to compile for wasm32.

**Comment by kinke** [comment]
This was fixed in druntime v2.102 (=> LDC v1.32).
